src/Compilers/CSharp/Portable/Compilation/CSharpSemanticModel.cs

```diff
@@ -2192,13 +2192,18 @@ private static bool IsUserDefinedTrueOrFalse(BoundUnaryOperator @operator)
                         convertedNullability = convertedCollection.TopLevelNullability;
                         conversion = convertedCollectionConversion;
                     }
+                    else if (highestBoundExpr is BoundConversion { ConversionKind: ConversionKind.ImplicitNullable, Conversion.UnderlyingConversions: [{ Kind: ConversionKind.CollectionExpression }] } boundConversion)
+                    {
+                        convertedType = highestBoundExpr.Type;
+                        convertedNullability = convertedCollection.TopLevelNullability;
+                        conversion = boundConversion.Conversion;
+                    }
                     else
                     {
-                        // There was an explicit cast on top of this.
-                        (convertedType, convertedNullability) = (convertedCollection.Type, nullability);
-                        conversion = convertedCollection.CollectionTypeKind == CollectionExpressionTypeKind.None
-                            ? Conversion.NoConversion
-                            : Conversion.CollectionExpression;
+                        // Explicit cast or error scenario like `object x = [];`
```

Comment:
>Explicit cast [](http://example.com/codeflow?start=27&length=13)

Does this case still cover explicit casts? (If so, it's surprising the converted type is `null` and the conversion is `Identity`.) If not, consider updating the comment.

Reply:
Yes, this case still covers explicit casts (test `NullableValueType_ExplicitCast` for instance).

The behavior for `ConvertedType` and the conversion matches what we have for other typeless expressions with an explicit cast:
```
        [Fact]
        public void TODO2()
        {
            string src = """
partial class Program
{
    static object M()
    {
        return (object)null;
    }
}
""";
            var comp = CreateCompilation(src, targetFramework: TargetFramework.Net80);
            comp.VerifyDiagnostics();

            var tree = comp.SyntaxTrees.First();
            var model = comp.GetSemanticModel(tree);

            var value = tree.GetRoot().DescendantNodes().OfType<CastExpressionSyntax>().Last().Expression;
            Assert.Equal("null", value.ToFullString());
            var conversion = model.GetConversion(value);
            Assert.True(conversion.IsIdentity);

            var typeInfo = model.GetTypeInfo(value);
            Assert.Null(typeInfo.Type);
            Assert.Null(typeInfo.ConvertedType);
        }
```